notebooks/02-ingestion/10-updating-data.ipynb

```diff
@@ -0,0 +1,1188 @@
+    "  CASE WHEN (t.\"__time\" IS NULL OR t.\"__time\" = s.\"__time\") THEN s.\"departure_delay\"\n",
+    "       ELSE t.\"departure_delay\" \n",
+    "  END AS \"departure_delay\"\n",
+    "FROM \"example-flights-updates\" t \n",
+    "  FULL OUTER JOIN\n",
+    "    \"example-upsert-changeset\" s\n",
+    "  ON t.\"__time\"=s.\"__time\"\n",
+    "  AND t.\"airline\"=s.\"airline\"\n",
+    "  AND t.\"flight_number\"=s.\"flight_number\" \n",
+    "WHERE t.\"__time\" IS NULL OR TIME_IN_INTERVAL(t.\"__time\", '2005-11-03T00:00:00.000Z/P1D') \n",
+    "PARTITIONED BY DAY    -- replacing a whole day so partition by day\n",
+    "'''\n",
+    "req = sql_client.sql_request(sql)\n",
+    "req.add_context(\"sqlJoinAlgorithm\", 'sortMerge')\n",
+    "display.run_task(req)\n",
+    "sql_client.wait_until_ready('example-flights-updates')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "faf24eea-956e-4531-8708-f2e4064757fa",
+   "metadata": {},
+   "source": [
+    "Check the data changes and new rows with the same SQL as before:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7b62872f-9ee6-4ab7-afdc-43d91c522215",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT min(\"flight_number\") as \"min_flight\", max(\"flight_number\") as \"max_flight\", \n",
+    "       count(*) \"total_flights_day\", \n",
+    "       SUM(\"departure_delay\") \"total_time_lost\" \n",
+    "FROM \"example-flights-updates\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\", '2005-11-03T00:00:00.000Z/P1D')  \n",
+    "  AND \"airline\"='TZ'\n",
+    "'''\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9721e4f0-a208-47e0-8dad-da86b36185f7",
+   "metadata": {},
+   "source": [
+    "Notice in the result above that:\n",
+    "- there are now 9 new rows\n",
+    "- that the `departure_delay` values have been updated producing the `total_time_lost` = 0.\n",
+    "\n",
+    "Just as a little extra fun:\n",
+    "At first the result seems surprising because some of the flights were left as is, the ones with flight numbers between 4901 amd 4999 were replaced with 5000, so those rows would not have been touched. It's an exercise for the student to verify it. I did verify it by checking that the flights in the 9 original unchanged rows all had zero delay values."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4078c260-7e34-452d-b234-12d8d19908c1",
+   "metadata": {},
+   "source": [
+    "## Revisionist history update\n",
+    "Imagine for example, that you have credit transaction activity data over a years of history and for a period of a few months a person's identity had been stolen. The person worked it all out with the banks and now there is new validated data for what their transactions were. So we need to do an update across a large portion of the timeline, removing the rows for this person and inserting the new ones across the same period of time.\n",
+    "\n",
```

Comment:
I think technically they would add a transaction to the ledger, reversing bogus transactions, rather than deleting them. You therefore need a different story here.

Reply:
I think they actually fix your credit history at the credit bureaus. The right to be forgotten is not an update of history though, it is a delete.... another story?